db/table_cache.cc

```diff
@@ -60,6 +60,24 @@ void AppendVarint64(IterKey* key, uint64_t v) {
   key->TrimAppend(key->Size(), buf, ptr - buf);
 }
 
+// Get the sequence number for row cache key
+uint64_t GetSequenceNumber(const ReadOptions& options, const FileDescriptor& fd,
+                           const Slice& internal_key, GetContext* get_context) {
+  uint64_t seq_no = 0;
+
+  // Maybe we can include the whole file ifsnapshot == fd.largest_seqno.
+  if (options.snapshot != nullptr &&
+      (get_context->has_callback() ||
+       static_cast_with_check<const SnapshotImpl>(options.snapshot)
+               ->GetSequenceNumber() <= fd.largest_seqno)) {
+    // We should consider to use options.snapshot->GetSequenceNumber()
+    // instead of GetInternalKeySeqno(k), which will make the code
+    // easier to understand.
+    seq_no = 1 + GetInternalKeySeqno(internal_key);
```

Comment:
When `options.snapshot != nullptr`, the row cache entry will store `1 + options.snapshot` as sequence number. When this sequence number is provided to `replayGetContextLog` and later `SaveValue`, the registered callback as you linked https://github.com/facebook/rocksdb/blob/8fc78a3a9e1d24ba55731b70c0c25cef0765dbc8/db/db_impl/db_impl.cc#L2112 would fail still since  `1 + options.snapshot`(ParsedInternalKey.sequence) > options.snapshot (max_visible_key). So I think this fix won't work for `options.snapshot != nullptr` case. Would you mind adding a test case to check that?

Reply:
Thanks for pointing out. Test scenarios on getting row cache with both timestamp and snapshot being set are added.

Providing the `seq_no` of cache entry key is set to `0` by default, and has value `1 + snapshot's seq_no` if snapshot being provided, I think we can use `seq_no` for managing visibility of row cache entry and pass the value `seq_no - 1`(for non-zero seq_no) to build the internal key when row cache hits.